### ISSUE TITLE
Typescript extractions broken in next (2) version

### ISSUE THREAD

**Issue by haikyuu** [rematch/rematch#797]
Note I'm using the `2.0.0-next.1` version of rematch.

**Describe the bug**
Duplicate of https://github.com/rematch/rematch/issues/787
This bug has been reported before and a fix was suggested here for instance https://github.com/rematch/rematch/pull/788. But it was closed as already fixed. And it seems it's still an issue.

**Expected behavior**
Dispatch and state types should be extracted correctly.

**Screenshots**

![image](https://user-images.githubusercontent.com/8558836/89933027-5a3bbb80-dc06-11ea-89b6-4b9edd8d6823.png)
![image](https://user-images.githubusercontent.com/8558836/89933053-6293f680-dc06-11ea-9403-2b422b4894df.png)
![image](https://user-images.githubusercontent.com/8558836/89933089-72133f80-dc06-11ea-97c5-a842e7d7b256.png)
![image](https://user-images.githubusercontent.com/8558836/89933276-ba326200-dc06-11ea-80d6-b3a0ba991590.png)

**Comment by semoal** [comment]
Could you provide us a reproduction repo? 

In this way we can debug better the issue

**Comment by haikyuu** [comment]
@semoal Sure. I'll create one

**Comment by haikyuu** [comment]
@semoal here it is. https://codesandbox.io/s/gifted-shape-6oywz?file=/src/models/questions.ts

**Comment by semoal** [comment]
Thanks @haikyuu we'll look asap onto it

**Comment by semoal** [comment]
If you need a workaround i've fixed it haikyuu, we'll look into extracter type of the state, because is the type that is failing.

Changed to this:

```js
  state: {
    questions: [] as Array<BooleanQuestion>,
    amount: 2,
    type: "boolean" as QuestionType
  }, // initial state
```

```
// state.d.ts
type QuestionType = "boolean" | "multiple" | "mixed";
```

**Comment by haikyuu** [comment]
@semoal Thanks a lot @semoal

**Comment by haikyuu** [comment]
@semoal 
> we'll look into extracter type of the state, because is the type that is failing.

Note that `Dispatch` types are also failing

**Comment by haikyuu** [comment]
@semoal When declaring the initial state and typing it in the upper scope, typing works just fine. I've updated the codesandbox and it works fine now.

It's probably an issue of type extractions needing types to be explicitly defined rather than inferred. And that's fine IMHO.

Another thing is that the user needs to type reducers and effects parameters to get their typings in other places in the app.

In that case, we just need to update the docs.

**Comment by haikyuu** [comment]
@semoal Here is a PR for the docs improvements https://github.com/rematch/rematch/pull/799

**Comment by semoal** [comment]
I've fixed the requirement of explicitly define the initial state, i'm fixing a minor issue on the dispatch function and i'll release a @next version with this issue fixed @haikyuu

**Comment by semoal** [comment]
Changed from this:
```typescript
export const createModel = <TState, TBaseState = TState>() => <
	TModel extends Model<TState, TBaseState>
>(
	model: TModel
): TModel => model
```
![Screenshot 2020-08-14 at 15 03 29](https://user-images.githubusercontent.com/22656541/90252180-59419e80-de3f-11ea-8a9e-80e1273973d0.png)

to this:
```typescript
export const createModel: <TState, TBaseState = TState>() => (
	mo: Model<TState, TBaseState>
) => Model<TState, TBaseState> = () => (model: Model): Model => model
```
![Screenshot 2020-08-14 at 15 04 12](https://user-images.githubusercontent.com/22656541/90252233-6d859b80-de3f-11ea-99d9-1c5ef5eb1b66.png)

Complex types looks fixed, but now the dispatcher function can't get it to work mate, any suggerence?
For me looks like the extractor methods are not working properly.
```
Property 'setQuestions' does not exist on type 'ModelDispatcher<Model<QuestionsState, QuestionsState>>'.
  Property 'setQuestions' does not exist on type 'undefined & void'.
```

@pjamrozowicz

**Comment by semoal** [comment]
🚀🚀🚀 Released the fix on 2.0.0-next.2 .🚀🚀🚀
Confirm @haikyuu that you're issue is fixed, you only need to things:
```js
createModel<State>
```
to
```js
 createModel<RootModel>
```

and in your model file:

```js
export RootModel extends Models<RootModel> {
...
}
```
